runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcGetDataStream.java

```diff
@@ -188,20 +195,32 @@ private long uniqueId() {
 
   @Override
   public KeyedGetDataResponse requestKeyedData(String computation, KeyedGetDataRequest request) {
-    return issueRequest(
-        QueuedRequest.forComputation(uniqueId(), computation, request),
-        KeyedGetDataResponse::parseFrom);
+    try {
+      return issueRequest(
+          QueuedRequest.forComputation(uniqueId(), computation, request),
+          KeyedGetDataResponse::parseFrom);
+    } catch (
+        org.apache.beam.runners.dataflow.worker.windmill.client.WindmillStreamShutdownException e) {
+      throw new WorkItemCancelledException(request.getShardingKey());
+    }
   }
 
   @Override
   public GlobalData requestGlobalData(GlobalDataRequest request) {
-    return issueRequest(QueuedRequest.global(uniqueId(), request), GlobalData::parseFrom);
+    try {
+      return issueRequest(QueuedRequest.global(uniqueId(), request), GlobalData::parseFrom);
+    } catch (
+        org.apache.beam.runners.dataflow.worker.windmill.client.WindmillStreamShutdownException e) {
+      throw new WorkItemCancelledException(
```

Comment:
this gets converted into GetDataException upstream, I think we need to remove GetDataException (could lead to repeated retries). Can be done in a separate PR.

Reply:
added TODO

Reply:
maybe better to not throw WorkItemCancellation here then and let the GetDataClient convert WindmillStreamException -> WorkItemCancellation